lsif-semanticdb/src/main/java/com/sourcegraph/lsif_semanticdb/SignatureFormatter.java

```diff
@@ -0,0 +1,309 @@
+package com.sourcegraph.lsif_semanticdb;
+
+import com.sourcegraph.semanticdb_javac.Semanticdb.SymbolInformation.Property;
+import com.sourcegraph.semanticdb_javac.Semanticdb.*;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+public class SignatureFormatter {
+  private static final Type OBJECT_TYPE_REF =
+      Type.newBuilder().setTypeRef(TypeRef.newBuilder().setSymbol("java/lang/Object#")).build();
+
+  private static final Type WILDCARD_TYPE_REF =
+      Type.newBuilder().setTypeRef(TypeRef.newBuilder().setSymbol("local_wildcard")).build();
+
+  private static final String ARRAY_SYMBOL = "scala/Array#";
+  private static final String ENUM_SYMBOL = "java/lang/Enum#";
+  private static final String ANNOTATION_SYMBOL = "java/lang/annotation/Annotation#";
+
+  private final StringBuilder s = new StringBuilder();
+  private final SymbolInformation symbolInformation;
+  private final Symtab symtab;
+
+  public SignatureFormatter(SymbolInformation symbolInformation, Symtab symtab) {
+    this.symbolInformation = symbolInformation;
+    this.symtab = symtab;
+  }
+
+  public String formatSymbol() {
+    Signature signature = symbolInformation.getSignature();
+    if (signature.hasClassSignature()) {
+      formatClassSignature(signature.getClassSignature());
+    } else if (signature.hasMethodSignature()) {
+      formatMethodSignature(signature.getMethodSignature());
+    } else if (signature.hasValueSignature()) {
+      formatValueSignature(signature.getValueSignature());
+    } else if (signature.hasTypeSignature()) {
+      formatTypeParameterSignature(signature.getTypeSignature());
+    }
+
+    return s.toString();
+  }
+
+  private void formatClassSignature(ClassSignature classSignature) {
+    boolean isAnnotation =
+        classSignature.getParentsList().stream()
+            .anyMatch(t -> t.getTypeRef().getSymbol().equals(ANNOTATION_SYMBOL));
+
+    printKeyword(formatAccess());
+    if (!has(Property.ENUM) && !isAnnotation) printKeyword(formatModifiers());
+
+    switch (symbolInformation.getKind()) {
+      case CLASS:
+        if (has(Property.ENUM)) {
+          printKeyword("enum");
+        } else {
+          printKeyword("class");
+        }
+        break;
+      case INTERFACE:
+        if (isAnnotation) {
+          printKeyword("@interface");
+          break;
+        }
+        printKeyword("interface");
+        break;
+    }
+    s.append(symbolInformation.getDisplayName());
+
+    List<SymbolInformation> typeParameters = getSymlinks(classSignature.getTypeParameters());
+    if (!typeParameters.isEmpty()) {
+      s.append(
+          typeParameters.stream()
+              .map(this::formatTypeParameter)
+              .collect(Collectors.joining(", ", "<", ">")));
+    }
+
+    List<Type> parentsList =
```

Comment:
nit: the name "parentsList" would hint this is just normal parents 
```suggestion
    List<Type> nonSyntheticParents =
```